### ISSUE TITLE
OpenGL views don't display for some Windows users

### ISSUE THREAD

**Issue by bhaller** [MesserLab/SLiM#462]
From a slim-discuss thread at https://groups.google.com/d/msgid/slim-discuss/c149277d-cff7-4956-8238-b9e5fc0f96e5n%40googlegroups.com?utm_medium=email&utm_source=footer :

> upon installing SLiM and the SLiMgui on windows10 everything seems to work fine, except for the empty population and chromosome view windows. Has this problem been encountered before? The lettering in the gui was also small, but I was able to fix that with the suggestions provided in the manual.

A screenshot was provided:

![simgui_problem](https://github.com/user-attachments/assets/bf8ff37d-fe77-4af8-9a08-59da19f3c557)

Another user wrote:

> Dear Ben, it seems I have the same problem in SLIM on Windows 10. I don't see anything in chromosome and population windows. And when I try to create haplotype plot, it is also empty.

My reply was:

> Hmm.  Well, those two views are drawn using OpenGL, whereas all the other views in the window are not.  So something is wrong with Qt's use of OpenGL on your system. (Qt being the widget kit that SLiMgui is built on top of.)  Might be a driver incompatibility with your video card or something; I really can't even begin to guess.  If you can just run it on a different machine, that is likely to fix the problem, since nobody else has reported this issue.  You could also look for console logs from Qt related to this problem; maybe that will give a clue as to how you might fix it, if it gives error messages about OpenGL or about your video driver or something.  Update your video drivers, try a different video card if you can, etc.  Sorry I don't have a better answer.

I'm creating this issue to track this problem and provide a place for people to add comments, if anybody discovers a workaround or something.  I doubt there is anything I can do about it, short of migrating SLiM to Qt 6, which might or might not fix the problem.  If there is a Windows developer out there who has ideas for how they might debug/fix this issue, help would be great.

**Comment by alexpiper** [comment]
I believe im having the same/similar issue on windows with SLiM 4.2.2. At initialisation and tick1 of similulation there is a red and grey square in the individual view and the chromosome view is blank.

![image](https://github.com/user-attachments/assets/6b461184-4733-44bf-aecd-9a9a11db8697)

Then from tick 2 onwards both views are blank. Similar to the previous user the haplotype plot is also blank.

![image](https://github.com/user-attachments/assets/d69070ae-f924-4e2f-a163-653f7c96a163)

I have tried updating both the integrated graphics and Nvidia graphics card drivers with no success

**Comment by alexpiper** [comment]
I'm by no means an expert with compiling on windows. But definitely happy to try and help with testing any changes that might resolve the issue. Ive got a sysadmin in the building i can pester too if i get stuck.

**Comment by bhaller** [comment]
Great, thanks @alexpiper.  I'm totally swamped with other things at the present moment, but I hope to get to this soon.

**Comment by bhaller** [comment]
OK, progress.  The GitHub head version of SLiM can now build SLiMgui against Qt6 as well as Qt5, and my hope is that Qt6 will just magically fix this problem.  CMake will automatically detect Qt6, and prefer it to Qt5, so you can leave Qt5 installed if you wish, and just additionally install Qt6.  Then build SLiMgui from current GitHub sources, run it, check the About panel to confirm that it was built against Qt6 (it says the Qt version in that panel), see if the bug is fixed, and report back here.  @alexpiper, it would be great if you could try this out.  If anybody else is also lurking here who could try it, that would be very helpful.  Thanks!

**Comment by DanPell** [comment]
Hi, I'm experiencing the same issue as @alexpiper described above. As recommended by @bhaller I'm trying to build slim from the current head sources on GitHub, following the instructions in section 2.3.2 of the SLiM manual, but currently getting an error at the "build" step:

```
$ make
[  1%] Performing build step for 'libgnu'
sh: line 1: autom4te: command not found
aclocal-1.16: error: autom4te failed with exit status: 127
WARNING: 'aclocal-1.16' is missing on your system.
         You should only need it if you modified 'acinclude.m4' or
         'configure.ac' or m4 files included by 'configure.ac'.
         The 'aclocal' program is part of the GNU Automake package:
         <https://www.gnu.org/software/automake>
         It also requires GNU Autoconf, GNU m4 and Perl in order to run:
         <https://www.gnu.org/software/autoconf>
         <https://www.gnu.org/software/m4/>
         <https://www.perl.org/>
make[3]: *** [Makefile:979: /C/slim/SLiM-master/SLiM-master/windows_compat/gnulib/aclocal.m4] Error 127
make[2]: *** [CMakeFiles/libgnu.dir/build.make:86: libgnu-prefix/src/libgnu-stamp/libgnu-build] Error 2
make[1]: *** [CMakeFiles/Makefile2:95: CMakeFiles/libgnu.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

I'm new to using msys2 so I'm not sure what's wrong and would greatly appreciate advice! I've installed automake 1.16 as per the msys2 website instructions but that hasn't helped.
Thanks!

**Comment by bhaller** [comment]
@rdinnager any idea what's happening here?  I don't see any mention of automake/autoconf in the CMakeLists.txt file, so I'm a bit mystified as to why it would be involved here.  @DanPell you did already perform the steps in section 2.3., yes?  Installing prerequisites, running `cmake` before `make`, etc.?

**Comment by DanPell** [comment]
Yes, I've been following section 2.3.2, so:
```
pacman -Syu
pacman -S mingw-w64-x86_64-cmake
mkdir build
cd build
cmake -G"MSYS Makefiles" ..
make
```
Thanks

**Comment by bhaller** [comment]
@samchamper maybe you have a clue on this?

**Comment by alexpiper** [comment]
Chiming in to say that i had the same error as @DanPell while compiling. I got past it by installing autotools ` pacman -S mingw-w64-x86_64-autotools` It then compiled successfully, but with QT 5 still. 

![image](https://github.com/user-attachments/assets/161a38f6-7824-49ab-84f8-643e2bf33b80)

I did install qt6 in advance of compiling, but ill try again, perhaps removing qt5 first, and update you on how it goes.

**Comment by bhaller** [comment]
> Chiming in to say that i had the same error as @DanPell while compiling. I got past it by installing autotools ` pacman -S mingw-w64-x86_64-autotools` 

Interesting.  I wonder why that is now required, when it wasn't before?  Perhaps they dropped it from the list of packages included, or something.  Well, thanks for figuring it out; I'll add this to the standard Windows build instructions!

> It then compiled successfully, but with QT 5 still.
> 
> I did install qt6 in advance of compiling, but ill try again, perhaps removing qt5 first, and update you on how it goes.

Hmm.  Interesting.  CMake should be preferring Qt6, if both are installed.  Perhaps it is having trouble finding Qt6; or perhaps the preference for Qt6 is not working properly.  I'll be interested to hear more, thanks.

**Comment by rdinnager** [comment]
The `autotools` thing sounds vaguely familiar. I might have run into an issue that was solved in this way in the past, but unfortunately forgot to make note of it. Glad to see it added to the manual! I'd be interested in trying my hand at compiling with Qt6 under Windows myself, to help figure out how to make it work. Unfortunately I am teaching a course that starts next Monday, so I am completely taken up by prep at the moment. If I have some time next week I might give it a try. In the meantime I'm also interested in any progress you might make @DanPell. Thanks for your persistence so far!

**Comment by alexpiper** [comment]
After removing Qt5 it compiled with Qt6, with both present it defaulted to Qt5.

Unfortunately though, moving to Qt6 doesnt seem to have fixed the issue:
![image](https://github.com/user-attachments/assets/449fff9d-65cd-421b-87e4-83d7d6ec376c)

![image](https://github.com/user-attachments/assets/02474c17-5027-4b9f-afa3-d99484d76dad)

**Comment by bhaller** [comment]
@alexpiper very odd.  A couple of questions.  Do you have more than one screen attached to this machine, i.e., a multimonitor setup?  If you detach secondary screens to make it a single-screen machine (and then relaunch SLiMgui), does anything change?  If you drag the window from one screen to another, does anything change?  If you resize the window, does anything change?  And you have tried recycling and playing the model; this display never changes?  You wrote before, "from tick 2 onwards both views are blank"; this remains the case with Qt6?

The red pattern, by the way, is because the view has internally resized itself because it finds itself on a high-DPI screen; it has doubled its number of pixels.  Or possibly the opposite: it has double the number of pixels, because it is on a high-DPI screen, but it has failed to internally resize itself, and so most of it displays in red.  Either way, it ought to be a temporary state that would fix itself if you start playing the model, or resize the window, or things like that that trigger a redisplay of the view.  Apparently something is preventing it from redisplaying correctly.

> After removing Qt5 it compiled with Qt6, with both present it defaulted to Qt5.

Hmm, OK; I'll have to figure out why that is happening, too...

**Comment by bhaller** [comment]
> ... It then compiled successfully, but with QT 5 still... I did install qt6 in advance of compiling, but ill try again, perhaps removing qt5 first, and update you on how it goes.

> After removing Qt5 it compiled with Qt6, with both present it defaulted to Qt5.

Hi @alexpiper.  I hope that I have just fixed this problem.  You'll need to check out the latest GitHub head.  I'd apprciate knowing whether it now chooses Qt6 when both are installed.  Thanks a bunch for testing!

As for the main bug, see above.  I'll need to think of a new angle of attack on this problem.

**Comment by bhaller** [comment]
And if one of you two could test for this bug: https://github.com/MesserLab/SLiM/issues/461, that would be greatly appreciated.  So far I don't know whether it is correlated with this bug or not.  I'm hoping that bug, #461, gets fixed by Qt6, but maybe it doesn't either.  Frustrating not being able to test these problems myself.

**Comment by DanPell** [comment]
With the ```pacman -S mingw-w64-x86_64-autotools``` install, I was able to build slim from the github source- thanks! For the slimgui install, I replaced the ```pacman -S mingw-w64-x86_64-qt5-base``` command with ```pacman -S mingw-w64-x86_64-qt6-base```; is that the correct way of doing it with qt6? Probably not since when I try to open slimgui I get system errors stating that Qt6Gui.dll, QT6Core.dll, Qt6OpenGLWidgets.dll, Qt6Widgets.dll were not found.
As for bug #461, I've never experienced that, so unfortunately I can't test anything.
Thanks for the help!

**Comment by bhaller** [comment]
> With the `pacman -S mingw-w64-x86_64-autotools` install, I was able to build slim from the github source- thanks! For the slimgui install, I replaced the `pacman -S mingw-w64-x86_64-qt5-base` command with `pacman -S mingw-w64-x86_64-qt6-base`; is that the correct way of doing it with qt6? Probably not since when I try to open slimgui I get system errors stating that Qt6Gui.dll, QT6Core.dll, Qt6OpenGLWidgets.dll, Qt6Widgets.dll were not found. As for bug #461, I've never experienced that, so unfortunately I can't test anything. Thanks for the help!

Hmm, so you installed Qt6 with pacman (that seems fine to me?), then you built SLiMgui using CMake and the build was successful, but then when you tried to run it you got DLL errors?  Interesting.  I'm not sure why that would happen.  It might be that `qt6-base` is not the right package, maybe it doesn't include those components (but that would be weird in my opinion).  @alexpiper do you have advice on this?  @rdinnager will be great when you have time; I'm hoping to release SLiM 4.3 based on Qt6 pretty soon, but these build issues need to get sorted out first.

Good to know you don't experience #461.  It remains a mystery, but at least I know it isn't biting _everybody_.

**Comment by DanPell** [comment]
@bhaller I tried it with a few different external monitors, as well as just on my laptop, unfortunately it didn't change anything.
-Daniel

**Comment by alexpiper** [comment]
@bhaller I can confirm that with the latest git head its now compiling with Qt6 even if Qt5 is also installed.

In regards to the red box, similar to my original post from tick 2 onwards both views become blank with Qt6 and Qt5. I do have a multi monitor setup, but there is no difference if i use only a single monitor and relaunch SLiMgui, or resize the windows etc. 

I cant reproduce https://github.com/MesserLab/SLiM/issues/461 all the pop-up graphs work fine for me (with exception of the haplotype graph) and i can interact with and close them.

@DanPell i installed Qt6 exactly the same as you `pacman -S mingw-w64-x86_64-qt6-base` and didnt have any dll issues, so cant offer any help there unfortunately.

**Comment by bhaller** [comment]
OK, thanks folks.  Lots of useful info, but I have no idea how to proceed from here.  I wish I knew what the determining factor is – why certain people experience this problem while most don't.  Video hardware, video drivers, OS version, OpenGL version, something else.  But even if I knew the cause, it's not clear how I would then use that information to fix the problem; as far as I can tell, the issue is entirely within Qt or perhaps a lower level of the stack, not something I'm likely to have any control over in SLiM's code.

Well, I'm contemplating a possible workaround, which is to provide a checkbox in the Preferences panel: "Don't use OpenGL".  If checked, SLiMgui would switch to drawing the individuals view and the chromosome view without using OpenGL; it would be substantially slower (especially with many individuals / many mutations), but it would work.  Stay tuned.

For now, good to know for sure that it is unrelated to #461, and that Qt6 definitely does not fix the problem.

**Comment by bhaller** [comment]
Another idea is that there might be additional packages you need to install to obtain OpenGL on your system, for Qt's purposes.  I have no idea, but perhaps Google would lead you to a productive path.  It seems conceivable to me that Qt would basically disable OpenGL drawing if it can't find the necessary packages/drivers, but otherwise continue to run (so SLiMgui continues to work otherwise).  Have you guys tried launching SLiMgui from your terminal app, rather than double-clicking it?  You might see some useful error output about such a state, written to the terminal by Qt.  Or perhaps there is a "console" app you can check for error logs, even when you double-click the app?  (There is on macOS, I know.)  I rather like this idea.  Perhaps for many/most users these OpenGL resources have already been installed, by other software's installers for example, but on your machines it happens not to be present.

**Comment by alexpiper** [comment]
I agree, there is a bit more debugging i could do on my side with openGL and console. Ive also got an older windows PC in the office that worked fine with SLiMgui (an older version) in the past, and i can try look into what might be different between the two machines. Ill get onto this next week.

**Comment by bhaller** [comment]
Hey folks!  Well, the underlying problem remains a mystery, but for situations where this problem comes up, I have now added a "Use OpenGL" checkbox in the Preferences panel.  If you uncheck that checkbox, SLiMgui will no longer use OpenGL at all, and that should (I hope) fix the problem.  You will need to check out the latest version from GitHub and build it afresh.  No other changes needed.  @alexpiper @DanPell it would be great to hear if this does, in fact, fix the problem for you.  If not, there's a way I can take this one step further.  Let me know, thanks!

**Comment by bhaller** [comment]
Hey folks.  Working on putting together a SLiM 4.3 release; I'd really like to know whether the new "Use OpenGL" checkbox fixing this problem for you guys.  Thanks!  @alexpiper @DanPell

**Comment by alexpiper** [comment]
Confirming that unchecking the "Use OpenGL for fast display" resolves the issue for me!

**Comment by bhaller** [comment]
Thanks!  Given that, I'm going to close this issue.  I'm pretty certain the underlying issue is a bug in Qt, a bug in a graphics device driver, a bug in the OS, etc., rather than a bug in SLiMgui itself, so I think this workaround is the best we're going to do.  And I think it's a fine workaround, so I'm not unhappy with this outcome.

**Comment by DanPell** [comment]
> Hey folks! Well, the underlying problem remains a mystery, but for situations where this problem comes up, I have now added a "Use OpenGL" checkbox in the Preferences panel. If you uncheck that checkbox, SLiMgui will no longer use OpenGL at all, and that should (I hope) fix the problem. You will need to check out the latest version from GitHub and build it afresh. No other changes needed. @alexpiper @DanPell it would be great to hear if this does, in fact, fix the problem for you. If not, there's a way I can take this one step further. Let me know, thanks!

Hi Ben, sorry I didn't see this until now- I'd like to try this fix but unfortunately I haven't had time to debug my issues with building SLiM from the github source :( if I've got more time at some point I'll try again but I'm not experienced enough to know what I'm doing

**Comment by bhaller** [comment]
> Hi Ben, sorry I didn't see this until now- I'd like to try this fix but unfortunately I haven't had time to debug my issues with building SLiM from the github source :( if I've got more time at some point I'll try again but I'm not experienced enough to know what I'm doing

Hi @DanPell!  No worries.  Somebody else was able to test the fix and confirm it; I think it is fine.  You'll get the fix whenever you update to SLiM 4.3, which should be released quite soon now.  Thanks!
